src/pixie/paths.nim

```diff
@@ -1233,43 +1234,122 @@ proc fillShapes(
+      a = vec2(
+        at.x + normal.x * halfStroke,
+        at.y - normal.y * halfStroke
+      )
+      b = vec2(
+        to.x + normal.x * halfStroke,
+        to.y - normal.y * halfStroke
+      )
+      c = vec2(
+        to.x - normal.x * halfStroke,
+        to.y + normal.y * halfStroke
+      )
+      d = vec2(
+        at.x - normal.x * halfStroke,
+        at.y + normal.y * halfStroke
+      )
+
+    @[a, b, c, d, a]
 
   for shape in shapes:
-    var
-      strokeShape: seq[Vec2]
-      back: seq[Vec2]
-    for segment in shape.segments:
+    var shapeStroke: seq[seq[Vec2]]
+
+    proc makeJoin(prevPos, pos, nextPos: Vec2): seq[Vec2] =
+      let angle = fixAngle(-angle(prevPos - pos) - -angle(nextPos - pos))
+      if abs(abs(angle) - PI) > epsilon:
+        var
+          a = (pos - prevPos).normalize() * halfStroke
+          b = (pos - nextPos).normalize() * halfStroke
+        if angle >= 0:
+          a = vec2(-a.y, a.x)
+          b = vec2(b.y, -b.x)
+        else:
+          a = vec2(a.y, -a.x)
+          b = vec2(-b.y, b.x)
+
+        var lineJoin = lineJoin
+        if lineJoin == ljMiter and abs(angle) < miterAngleLimit:
+          lineJoin = ljBevel
+
+        case lineJoin:
+        of ljMiter:
+          let
+            la = line(prevPos + a, pos + a)
+            lb = line(nextPos + b, pos + b)
+          var at: Vec2
+          if la.intersects(lb, at):
+            return @[pos + a, at, pos + b, pos, pos + a]
+
+        of ljBevel:
+          return @[a + pos, b + pos, pos, a + pos]
+
+        of ljRound:
+          return makeCircle(prevPos)
+
+    if shape[0] != shape[^1]:
+      # This shape does not end at the same point it starts so draw the
+      # first line cap.
+      case lineCap:
+      of lcButt:
+        discard
+      of lcRound:
+        shapeStroke.add(makeCircle(shape[0]))
+      of lcSquare:
+        let tangent = (shape[1] - shape[0]).normalize()
+        shapeStroke.add(makeRect(
+          shape[0] - tangent * halfStroke,
+          shape[0]
+        ))
+
+    for i in 1 ..< shape.len:
       let
-        tangent = (segment.at - segment.to).normalize()
-        normal = vec2(-tangent.y, tangent.x)
-        left = segment(
-          segment.at - normal * widthLeft,
-          segment.to - normal * widthLeft
-        )
-        right = segment(
-          segment.at + normal * widthRight,
-          segment.to + normal * widthRight
-        )
+        pos = shape[i]
+        prevPos = shape[i - 1]
 
-      strokeShape.add([right.at, right.to])
-      back.add([left.at, left.to])
+      shapeStroke.add(makeRect(prevPos, pos))
 
-    # Add the back side reversed
-    for i in 1 .. back.len:
-      strokeShape.add(back[^i])
+      # If we need a line join
+      if i < shape.len - 1:
+        shapeStroke.add(makeJoin(prevPos, pos, shape[i + 1]))
 
-    strokeShape.add(strokeShape[0])
+    if shape[0] == shape[^1]:
+      shapeStroke.add(makeJoin(shape[^2], shape[^1], shape[1]))
```

Comment:
Is it possible to some how have a shape.len < 2? We should check for that? what does [(0,0), (0,0)] shape do?

Reply:
In the [(0,0), (0,0)] case (and others), horizontal lines are removed so it would have shape len of 0 and not do any stroking. Added tests for these types of scenarios.